node/repo/blockstore_opts.go

```diff
@@ -0,0 +1,52 @@
+package repo
+
+import badgerbs "github.com/filecoin-project/lotus/lib/blockstore/badger"
+
+// BadgerBlockstoreOptions returns the badger options to apply for the provided
+// domain.
+func BadgerBlockstoreOptions(domain BlockstoreDomain, path string, readonly bool) (badgerbs.Options, error) {
+	if domain != BlockstoreChain {
+		return badgerbs.Options{}, ErrInvalidBlockstoreDomain
+	}
+
+	opts := badgerbs.DefaultOptions(path)
+
+	// Due to legacy usage of blockstore.Blockstore, over a datastore, all
+	// blocks are prefixed with this namespace. In the future, this can go away,
+	// in order to shorten keys, but it'll require a migration.
+	opts.Prefix = "/blocks/"
+
+	// Blockstore values are immutable; therefore we do not expect any
+	// conflicts to emerge.
+	opts.DetectConflicts = false
+
+	// This is to optimize the database on close so it can be opened
+	// read-only and efficiently queried. We don't do that and hanging on
+	// stop isn't nice.
+	opts.CompactL0OnClose = false
+
+	// The alternative is "crash on start and tell the user to fix it". This
+	// will truncate corrupt and unsynced data, which we don't guarantee to
+	// persist anyways.
+	opts.Truncate = true
+
+	// We mmap the index into memory, and access values from disk.
+	// Ideally the table loading mode would be settable by LSM level.
+	opts.ValueLogLoadingMode = badgerbs.FileIO
+	opts.TableLoadingMode = badgerbs.MemoryMap
+
+	// Embed only values < 128 bytes in the LSM; larger values in value logs.
+	opts.ValueThreshold = 128
+
+	// Reduce this from 64MiB to 16MiB. That means badger will hold on to
+	// 20MiB by default instead of 80MiB. This does not appear to have a
+	// significant performance hit.
+	opts.MaxTableSize = 16 << 20
```

Comment:
The issue with this is the number of TableFiles that will be created. Lotus for a while now run with default from badger (64M) instead of 20M because of that.

Reply:
Our badger2 datastore adapter set this value here:

https://github.com/ipfs/go-ds-badger2/blob/a4499bb2f2431c7c430d8a5ffe65c69339e52a00/datastore.go#L107

But it turns out we take those defaults and then entirely override them here:

https://github.com/filecoin-project/lotus/blob/6cdbf5216e8e8101529cc67bb4b6468a886e1bc3/node/repo/fsrepo_ds.go#L23

This also means that we lose other defaults that made sense, e.g. ValueLogLoadingMode = FileIO instead of mmap, because we reverted to badger defaults:

https://pkg.go.dev/github.com/dgraph-io/badger#Options.WithValueLogLoadingMode

This could explain some things...

Reply:
Restored the MaxTableSize value here: 0b8a21e.

Reply:
I didn't find the `ValueLogLoadingMode = FileIO` having any benefit (and the results that suggested it are super old) so unless you have some results confirminig it, I would stay with dgraph/badger defaults.

Reply:
`FileIO` sabotages the benefits of `View()`, so we're back to mmap.